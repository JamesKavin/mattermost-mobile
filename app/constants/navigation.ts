// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import keyMirror from '@utils/key_mirror';

const Navigation = keyMirror({
    NAVIGATE_TO_TAB: null,
    NAVIGATION_HOME: null,
    NAVIGATION_SHOW_OVERLAY: null,
});

export default Navigation;
