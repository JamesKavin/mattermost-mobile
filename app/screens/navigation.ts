// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Appearance, DeviceEventEmitter, Keyboard, Platform} from 'react-native';
import {Layout, Navigation, Options, OptionsModalPresentationStyle} from 'react-native-navigation';
import merge from 'deepmerge';

import {Navigation as NavigationConstants, Preferences, Screens} from '@constants';

import EphemeralStore from '@store/ephemeral_store';

import type {LaunchProps} from '@typings/launch';

function getThemeFromState() {
    if (Appearance.getColorScheme() === 'dark') {
        return Preferences.THEMES.windows10;
    }
    return Preferences.THEMES.default;
}

export function resetToChannel(passProps = {}) {
    const theme = getThemeFromState();

    EphemeralStore.clearNavigationComponents();

    const stack = {
        children: [{
            component: {
                id: Screens.CHANNEL,
                name: Screens.CHANNEL,
                passProps,
                options: {
                    layout: {
                        componentBackgroundColor: theme.centerChannelBg,
                    },
                    statusBar: {
                        visible: true,
                    },
                    topBar: {
                        visible: false,
                        height: 0,
                        background: {
                            color: theme.sidebarHeaderBg,
                        },
                        backButton: {
                            visible: false,
                            color: theme.sidebarHeaderTextColor,
                        },
                    },
                },
            },
        }],
    };

    let platformStack: Layout = {stack};
    if (Platform.OS === 'android') {
        platformStack = {
            sideMenu: {
                left: {
                    component: {
                        id: Screens.MAIN_SIDEBAR,
                        name: Screens.MAIN_SIDEBAR,
                    },
                },
                center: {
                    stack,
                },
                right: {
                    component: {
                        id: Screens.SETTINGS_SIDEBAR,
                        name: Screens.SETTINGS_SIDEBAR,
                    },
                },
            },
        };
    }

    Navigation.setRoot({
        root: {
            ...platformStack,
        },
    });
}

export function resetToSelectServer(passProps: LaunchProps) {
    const theme = getThemeFromState();

    EphemeralStore.clearNavigationComponents();

    const children = [{
        component: {
            id: Screens.SERVER,
            name: Screens.SERVER,
            passProps: {
                ...passProps,
                theme,
            },
            options: {
                layout: {
                    backgroundColor: theme.centerChannelBg,
                    componentBackgroundColor: theme.centerChannelBg,
                },
                statusBar: {
                    visible: true,
                },
                topBar: {
                    backButton: {
                        color: theme.sidebarHeaderTextColor,
                        title: '',
                    },
                    background: {
                        color: theme.sidebarHeaderBg,
                    },
                    visible: false,
                    height: 0,
                },
            },
        },
    }];

    Navigation.setRoot({
        root: {
            stack: {
                children,
            },
        },
    });
}

export function resetToTeams(name: string, title: string, passProps = {}, options = {}) {
    const theme = getThemeFromState();
    const defaultOptions = {
        layout: {
            componentBackgroundColor: theme.centerChannelBg,
        },
        statusBar: {
            visible: true,
        },
        topBar: {
            visible: true,
            title: {
                color: theme.sidebarHeaderTextColor,
                text: title,
            },
            backButton: {
                color: theme.sidebarHeaderTextColor,
                title: '',
            },
            background: {
                color: theme.sidebarHeaderBg,
            },
        },
    };

    EphemeralStore.clearNavigationComponents();

    Navigation.setRoot({
        root: {
            stack: {
                children: [{
                    component: {
                        id: name,
                        name,
                        passProps,
                        options: merge(defaultOptions, options),
                    },
                }],
            },
        },
    });
}

export function goToScreen(name: string, title: string, passProps = {}, options = {}) {
    const theme = getThemeFromState();
    const componentId = EphemeralStore.getNavigationTopComponentId();
    const defaultOptions = {
        layout: {
            componentBackgroundColor: theme.centerChannelBg,
        },
        popGesture: true,
        sideMenu: {
            left: {enabled: false},
            right: {enabled: false},
        },
        topBar: {
            animate: true,
            visible: true,
            backButton: {
                color: theme.sidebarHeaderTextColor,
                title: '',
                testID: 'screen.back.button',
            },
            background: {
                color: theme.sidebarHeaderBg,
            },
            title: {
                color: theme.sidebarHeaderTextColor,
                text: title,
            },
        },
    };

    Navigation.push(componentId, {
        component: {
            id: name,
            name,
            passProps,
            options: merge(defaultOptions, options),
        },
    });
}

export function popTopScreen(screenId?: string) {
    if (screenId) {
        Navigation.pop(screenId);
    } else {
        const componentId = EphemeralStore.getNavigationTopComponentId();
        Navigation.pop(componentId);
    }
}

export async function popToRoot() {
    const componentId = EphemeralStore.getNavigationTopComponentId();

    try {
        await Navigation.popToRoot(componentId);
    } catch (error) {
        // RNN returns a promise rejection if there are no screens
        // atop the root screen to pop. We'll do nothing in this case.
    }
}

export async function dismissAllModalsAndPopToRoot() {
    await dismissAllModals();
    await popToRoot();

    DeviceEventEmitter.emit(NavigationConstants.NAVIGATION_DISMISS_AND_POP_TO_ROOT);
}

export function showModal(name: string, title: string, passProps = {}, options = {}) {
    const theme = getThemeFromState();
    const modalPresentationStyle: OptionsModalPresentationStyle = Platform.OS === 'ios' ? OptionsModalPresentationStyle.pageSheet : OptionsModalPresentationStyle.none;
    const defaultOptions: Options = {
        modalPresentationStyle,
        layout: {
            componentBackgroundColor: theme.centerChannelBg,
        },
        statusBar: {
            visible: true,
        },
        topBar: {
            animate: true,
            visible: true,
            backButton: {
                color: theme.sidebarHeaderTextColor,
                title: '',
            },
            background: {
                color: theme.sidebarHeaderBg,
            },
            title: {
                color: theme.sidebarHeaderTextColor,
                text: title,
            },
            leftButtonColor: theme.sidebarHeaderTextColor,
            rightButtonColor: theme.sidebarHeaderTextColor,
        },
    };

    EphemeralStore.addNavigationModal(name);
    Navigation.showModal({
        stack: {
            children: [{
                component: {
                    id: name,
                    name,
                    passProps,
                    options: merge(defaultOptions, options),
                },
            }],
        },
    });
}

export function showModalOverCurrentContext(name: string, passProps = {}, options = {}) {
    const title = '';
    const animationsEnabled = (Platform.OS === 'android').toString();
    const defaultOptions = {
        modalPresentationStyle: 'overCurrentContext',
        layout: {
            backgroundColor: 'transparent',
            componentBackgroundColor: 'transparent',
        },
        topBar: {
            visible: false,
            height: 0,
        },
        animations: {
            showModal: {
                waitForRender: true,
                enabled: animationsEnabled,
                alpha: {
                    from: 0,
                    to: 1,
                    duration: 250,
                },
            },
            dismissModal: {
                enabled: animationsEnabled,
                alpha: {
                    from: 1,
                    to: 0,
                    duration: 250,
                },
            },
        },
    };
    const mergeOptions = merge(defaultOptions, options);

    showModal(name, title, passProps, mergeOptions);
}

export function showSearchModal(initialValue = '') {
    const name = 'Search';
    const title = '';
    const passProps = {initialValue};
    const options = {
        topBar: {
            visible: false,
            height: 0,
        },
        ...Platform.select({
            ios: {
                modalPresentationStyle: 'pageSheet',
            },
        }),
    };

    showModal(name, title, passProps, options);
}

export async function dismissModal(options = {}) {
    if (!EphemeralStore.hasModalsOpened()) {
        return;
    }

    const componentId = EphemeralStore.getNavigationTopComponentId();

    try {
        await Navigation.dismissModal(componentId, options);
        EphemeralStore.removeNavigationModal(componentId);
    } catch (error) {
        // RNN returns a promise rejection if there is no modal to
        // dismiss. We'll do nothing in this case.
    }
}

export async function dismissAllModals(options = {}) {
    if (!EphemeralStore.hasModalsOpened()) {
        return;
    }

    try {
        await Navigation.dismissAllModals(options);
        EphemeralStore.clearNavigationModals();
    } catch (error) {
        // RNN returns a promise rejection if there are no modals to
        // dismiss. We'll do nothing in this case.
    }
}

export function setButtons(componentId: string, buttons = {leftButtons: [], rightButtons: []}) {
    const options = {
        topBar: {
            ...buttons,
        },
    };

    mergeNavigationOptions(componentId, options);
}

export function mergeNavigationOptions(componentId: string, options: Options) {
    Navigation.mergeOptions(componentId, options);
}

export function showOverlay(name: string, passProps = {}, options = {}) {
    const defaultOptions = {
        layout: {
            backgroundColor: 'transparent',
            componentBackgroundColor: 'transparent',
        },
        overlay: {
            interceptTouchOutside: false,
        },
    };

    Navigation.showOverlay({
        component: {
            name,
            passProps,
            options: merge(defaultOptions, options),
        },
    });
}

export async function dismissOverlay(componentId: string) {
    try {
        await Navigation.dismissOverlay(componentId);
    } catch (error) {
        // RNN returns a promise rejection if there is no modal with
        // this componentId to dismiss. We'll do nothing in this case.
    }
}

export function openMainSideMenu() {
    if (Platform.OS === 'ios') {
        return;
    }

    const componentId = EphemeralStore.getNavigationTopComponentId();

    Keyboard.dismiss();
    Navigation.mergeOptions(componentId, {
        sideMenu: {
            left: {visible: true},
        },
    });
}

export function closeMainSideMenu() {
    if (Platform.OS === 'ios') {
        return;
    }

    Keyboard.dismiss();
    Navigation.mergeOptions(Screens.CHANNEL, {
        sideMenu: {
            left: {visible: false},
        },
    });
}

export function enableMainSideMenu(enabled: boolean, visible = true) {
    if (Platform.OS === 'ios') {
        return;
    }

    Navigation.mergeOptions(Screens.CHANNEL, {
        sideMenu: {
            left: {enabled, visible},
        },
    });
}

export function openSettingsSideMenu() {
    if (Platform.OS === 'ios') {
        return;
    }

    Keyboard.dismiss();
    Navigation.mergeOptions(Screens.CHANNEL, {
        sideMenu: {
            right: {visible: true},
        },
    });
}

export function closeSettingsSideMenu() {
    if (Platform.OS === 'ios') {
        return;
    }

    Keyboard.dismiss();
    Navigation.mergeOptions(Screens.CHANNEL, {
        sideMenu: {
            right: {visible: false},
        },
    });
}
