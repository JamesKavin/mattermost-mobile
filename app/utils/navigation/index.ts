// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {IntlShape} from 'react-intl';
import {Alert} from 'react-native';
import {Navigation, Options} from 'react-native-navigation';

import {Screens, ServerErrors} from '@constants';
import {isServerError} from '@utils/errors';

export const appearanceControlledScreens = new Set([
    Screens.ONBOARDING,
    Screens.SERVER,
    Screens.LOGIN,
    Screens.FORGOT_PASSWORD,
    Screens.MFA,
    Screens.SSO,
    Screens.REVIEW_APP,
    Screens.SHARE_FEEDBACK,
]);

export function mergeNavigationOptions(componentId: string, options: Options) {
    Navigation.mergeOptions(componentId, options);
}

export function alertTeamRemove(displayName: string, intl: IntlShape) {
    Alert.alert(
        intl.formatMessage({
            id: 'alert.removed_from_team.title',
            defaultMessage: 'Removed from team',
        }),
        intl.formatMessage({
            id: 'alert.removed_from_team.description',
            defaultMessage: 'You have been removed from team {displayName}.',
        }, {displayName}),
        [{
            style: 'cancel',
            text: intl.formatMessage({id: 'mobile.oauth.something_wrong.okButton', defaultMessage: 'OK'}),
        }],
    );
}

export function alertChannelRemove(displayName: string, intl: IntlShape) {
    Alert.alert(
        intl.formatMessage({
            id: 'alert.removed_from_channel.title',
            defaultMessage: 'Removed from channel',
        }),
        intl.formatMessage({
            id: 'alert.removed_from_channel.description',
            defaultMessage: 'You have been removed from channel {displayName}.',
        }, {displayName}),
        [{
            style: 'cancel',
            text: intl.formatMessage({id: 'mobile.oauth.something_wrong.okButton', defaultMessage: 'OK'}),
        }],
    );
}

export function alertChannelArchived(displayName: string, intl: IntlShape) {
    Alert.alert(
        intl.formatMessage({
            id: 'alert.channel_deleted.title',
            defaultMessage: 'Archived channel',
        }),
        intl.formatMessage({
            id: 'alert.channel_deleted.description',
            defaultMessage: 'The channel {displayName} has been archived.',
        }, {displayName}),
        [{
            style: 'cancel',
            text: intl.formatMessage({id: 'mobile.oauth.something_wrong.okButton', defaultMessage: 'OK'}),
        }],
    );
}

export function alertTeamAddError(error: unknown, intl: IntlShape) {
    let errMsg = intl.formatMessage({id: 'join_team.error.message', defaultMessage: 'There has been an error joining the team.'});

    if (isServerError(error)) {
        if (error.server_error_id === ServerErrors.TEAM_MEMBERSHIP_DENIAL_ERROR_ID) {
            errMsg = intl.formatMessage({
                id: 'join_team.error.group_error',
                defaultMessage: 'You need to be a member of a linked group to join this team.',
            });
        } else if (error.message) {
            errMsg = error.message;
        }
    }

    Alert.alert(
        intl.formatMessage({id: 'join_team.error.title', defaultMessage: 'Error joining a team'}),
        errMsg,
    );
}
