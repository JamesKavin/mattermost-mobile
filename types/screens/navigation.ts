// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {OptionsTopBarButton} from 'react-native-navigation/lib/src/interfaces/Options';

export type NavButtons = {
    leftButtons?: OptionsTopBarButton[];
    rightButtons?: OptionsTopBarButton[];
}
