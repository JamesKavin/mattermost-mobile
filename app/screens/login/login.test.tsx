// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {Preferences, Screens} from '@constants';
import * as NavigationActions from '@screens/navigation';
import {waitFor, renderWithIntl, fireEvent} from '@test/intl-test-helper';

import Login from './index';

jest.mock('@requests/remote/user', () => {
    return {
        login: () => {
            return {
                data: undefined,
                error: {
                    server_error_id: 'mfa.validate_token.authenticate.app_error',
                },
            };
        },
    };
});

describe('Login', () => {
    const baseProps = {
        componentId: Screens.LOGIN,
        config: {
            EnableSignInWithEmail: 'true',
            EnableSignInWithUsername: 'true',
        },
        license: {
            IsLicensed: 'false',
        },
        theme: Preferences.THEMES.default,
    };

    test('Login screen should match snapshot', () => {
        const {toJSON} = renderWithIntl(
            <Login {...baseProps}/>,
        );

        expect(toJSON()).toMatchSnapshot();
    });

    test('should show "I forgot my password" with only email login enabled', () => {
        const props = {
            ...baseProps,
            config: {
                ...baseProps.config,
                EnableSignInWithUsername: 'false',
            },
        };

        const {getByTestId} = renderWithIntl(<Login {...props}/>, {locale: 'es'});

        expect(getByTestId('login.forgot')).toBeDefined();
    });

    test('should show "I forgot my password" with only username login enabled', () => {
        const props = {
            ...baseProps,
            config: {
                ...baseProps.config,
                EnableSignInWithEmail: 'false',
            },
        };

        const {getByTestId} = renderWithIntl(<Login {...props}/>, {locale: 'fr'});

        expect(getByTestId('login.forgot')).toBeDefined();
    });

    test('should not show "I forgot my password" without email or username login enabled', () => {
        const props = {
            ...baseProps,
            config: {
                ...baseProps.config,
                EnableSignInWithEmail: 'false',
                EnableSignInWithUsername: 'false',
            },
        };

        const {getByTestId} = renderWithIntl(<Login {...props}/>);
        let forgot;

        try {
            forgot = getByTestId('login.forgot');
        } catch {
            // do nothing
        }

        expect(forgot).toBeUndefined();
    });

    test('should go to MFA screen when login response returns MFA error', async () => {
        const goToScreen = jest.spyOn(NavigationActions, 'goToScreen');

        const {getByTestId} = renderWithIntl(<Login {...baseProps}/>);
        const loginInput = getByTestId('login.username.input');
        const passwordInput = getByTestId('login.password.input');
        const loginButton = getByTestId('login.signin.button');
        const loginId = 'user';
        const password = 'password';

        fireEvent.changeText(loginInput, loginId);
        fireEvent.changeText(passwordInput, password);

        await waitFor(() => fireEvent.press(loginButton), {timeout: 300});

        expect(goToScreen).
            toHaveBeenCalledWith(
                'MFA',
                'Multi-factor Authentication',
                {
                    goToChannel: expect.anything(),
                    loginId,
                    password,
                    config: {EnableSignInWithEmail: 'true', EnableSignInWithUsername: 'true'},
                    license: {IsLicensed: 'false'},
                    theme: baseProps.theme,
                },
            );
    });

    test('should go to ForgotPassword screen when forgotPassword is called', () => {
        const goToScreen = jest.spyOn(NavigationActions, 'goToScreen');

        const {getByTestId} = renderWithIntl(<Login {...baseProps}/>);
        const forgot = getByTestId('login.forgot');

        fireEvent.press(forgot);

        expect(goToScreen).
            toHaveBeenCalledWith(
                'ForgotPassword',
                'Password Reset',
                {theme: baseProps.theme},
            );
    });
});
