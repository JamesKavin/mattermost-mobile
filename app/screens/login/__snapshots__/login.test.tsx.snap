// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login screen should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    accessible={false}
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingHorizontal": 15,
        "paddingVertical": 50,
      }
    }
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={true}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    focusable={true}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onScroll={[Function]}
    onStartShouldSetResponder={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "flex": 1,
      }
    }
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <Image
        source={
          Object {
            "testUri": "../../../dist/assets/images/logo.png",
          }
        }
        style={
          Object {
            "height": 72,
            "resizeMode": "contain",
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        blurOnSubmit={false}
        disableFullscreenUI={true}
        keyboardType="email-address"
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Email or Username"
        placeholderTextColor="rgba(61,60,64,0.5)"
        rejectResponderTermination={true}
        returnKeyType="next"
        style={
          Object {
            "alignSelf": "stretch",
            "borderColor": "gainsboro",
            "borderRadius": 3,
            "borderWidth": 1,
            "color": "#3d3c40",
            "fontSize": 16,
            "height": 45,
            "marginBottom": 5,
            "marginTop": 5,
            "paddingLeft": 10,
          }
        }
        testID="login.username.input"
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        disableFullscreenUI={true}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Password"
        placeholderTextColor="rgba(61,60,64,0.5)"
        rejectResponderTermination={true}
        returnKeyType="go"
        secureTextEntry={true}
        style={
          Object {
            "alignSelf": "stretch",
            "borderColor": "gainsboro",
            "borderRadius": 3,
            "borderWidth": 1,
            "color": "#3d3c40",
            "fontSize": 16,
            "height": 45,
            "marginBottom": 5,
            "marginTop": 5,
            "paddingLeft": 10,
          }
        }
        testID="login.password.input"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessibilityRole="button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#166de0",
            "borderRadius": 3,
            "borderWidth": 1,
            "marginTop": 10,
            "opacity": 1,
            "padding": 15,
          }
        }
        testID="login.signin.button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#166de0",
                "fontSize": 17,
                "textAlign": "center",
              },
              Object {},
            ]
          }
        >
          Sign in
        </Text>
      </View>
      <View
        accessibilityRole="button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "transparent",
            "marginTop": 15,
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2389d7",
            }
          }
          testID="login.forgot"
        >
          I forgot my password
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
